modules/indexing/src/test/java/org/apache/ignite/sqltests/CheckWarnJoinPartitionedTables.java

```diff
@@ -0,0 +1,408 @@
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        crd = startGrid();
+    }
+
+    /** */
+    @Test
+    public void joinSameTableWithPrimaryKey() {
+        execute(crd, new SqlFieldsQuery(
+            "CREATE TABLE A (ID INT PRIMARY KEY, TITLE VARCHAR);"));
+
+        checkSameTableWithJoinType("LEFT JOIN");
+        checkSameTableWithJoinType("RIGHT JOIN");
+        checkSameTableWithJoinType("INNER JOIN");
+        checkSameTableWithJoinType("JOIN");
+    }
+
+    /** */
+    @Test
+    public void joinSameTableWithPrimaryAffinityKey() {
+        execute(crd, new SqlFieldsQuery(
+            "CREATE TABLE A (ID INT PRIMARY KEY, TITLE VARCHAR) with \"AFFINITY_KEY=ID\";"));
+
+        checkSameTableWithJoinType("LEFT JOIN");
+        checkSameTableWithJoinType("RIGHT JOIN");
+        checkSameTableWithJoinType("INNER JOIN");
+        checkSameTableWithJoinType("JOIN");
+    }
+
+    /** */
+    private void checkSameTableWithJoinType(String joinType) {
+        checkLogListener(false,
+            "SELECT a1.* FROM A a1 " + joinType + " A a2 on a1.ID = a2.ID;");
+
+        checkLogListener(false,
+            "SELECT a1.* FROM A a1 " + joinType + " A a2 on a2.ID = a1.ID;");
+
+        checkLogListener(false,
+            "SELECT a1.* FROM A a1 " + joinType + " A a2 on a1.ID = a2._KEY;");
+
+        checkLogListener(false,
+            "SELECT a1.* FROM A a1 " + joinType + " A a2 on a1._KEY = a2.ID;");
+
+        checkLogListener(false,
+            "SELECT a1.* FROM A a1 " + joinType + " A a2 on a1._KEY = a2._KEY;");
+
+        checkLogListener(false,
+            "SELECT a1.* FROM A a1 " + joinType + " A a2 where a1.ID = a2.ID;");
+
+        checkLogListener(false,
+            "SELECT a1.* FROM A a1 " + joinType + " A a2 where a1.ID = a2._KEY;");
+
+        checkLogListener(false,
+            "SELECT a1.* FROM A a1 " + joinType + " A a2 on a1.ID = a2.ID AND a1.TITLE = a2.TITLE;");
+
+        checkLogListener(false,
+            "SELECT a1.* FROM A a1 " + joinType + " A a2 on a1.ID = a2.ID AND a1.TITLE != a2.TITLE;");
+
+        checkLogListener(false,
+            "SELECT a1.* FROM A a1 " + joinType + " A a2 on a1.ID = a2.ID AND (a1.TITLE = a2.TITLE OR a1.ID = a2.ID);");
+
+        checkLogListener(false,
+            "SELECT a1.* FROM A a1 " + joinType + " A a2 on (a1.TITLE = a2.TITLE OR a1.ID = a2.ID) AND a1.ID = a2.ID;");
+
+        checkLogListener(false,
+            "SELECT a1.* FROM A a1 " + joinType + " (select a2.ID from A a2) a3 on a1.ID = a3.ID;");
+
+        checkLogListener(false,
+            "SELECT a1.* FROM A a1 where a1.ID in (select a2.ID from A a2 " + joinType + " A a3 on a2.ID = a3.ID);");
+
+        checkLogListener(false,
+            "SELECT a1.* FROM A a1 " + joinType + " (select a2.ID from A a2 " + joinType + " A a3 on a2.ID = a3.ID) t" +
+                " on a1.ID = t.ID;");
+
+        // Ignite doesn't support this.
+        if (!joinType.contains("RIGHT"))
```

Comment:
Ignite doesn't support two RIGHT JOINs in the same statement or what?

Reply:
Yes, SQL engine write a message about unsupported clause.